### DISCUSSION TITLE
`i8` in `CodeAssembler`

### DISCUSSION THREAD

**Discussion by Kixiron** [icedland/iced#235] [Q&A] [answered]
I've noticed that `i8` aren't accepted for `imm8` operands and instead there's checks to see if the given operand of 32 or 64 bit size fits into 8 bits, is that on purpose? Additionally, unsigned integers don't have any trait implementations when they could

**Comment by Kixiron** [comment]
I'm gonna kinda use this as a store of disjoint thoughts so I don't bog things down with multiple issues/discussions:
- There's no way to get a `Register` from any of the `CodeAssembler` register types
- There's no way to "downcast" a register, like doing from `rax` to `al` or from `rbx` to `bl`. This'd be a really nice utility function

**Comment by wtfsck** [comment, marked as answer]
> I've noticed that `i8` aren't accepted for `imm8` operands and instead there's checks to see if the given operand of 32 or 64 bit size fits into 8 bits, is that on purpose?

To reduce the number of functions, it only accepts i32/u32 operands and sometimes i64/u64 operands.

> Additionally, unsigned integers don't have any trait implementations when they could

All functions that accept `i32` operands also accept `u32` operands. In other words, the same trait is implemented twice once for `i32` and once for `u32`.

>     * There's no way to get a `Register` from any of the `CodeAssembler` register types

Sounds like a feature request so it's better to create an issue.

Should be easy to fix, eg. `impl From<AsmRegister16> for Register` etc and maybe some `into_register()` methods.

>     * There's no way to "downcast" a register, like doing from `rax` to `al` or from `rbx` to `bl`. This'd be a really nice utility function

Should be an issue too.
